nrf-hal-common/src/usbd/mod.rs

```diff
@@ -0,0 +1,772 @@
+            regs.events_endepin[i].reset();
+            dma_end();
+
+            // Clear EPSTATUS.EPIN[i] flag
+            regs.epstatus.write(|w| unsafe { w.bits(1 << i) });
+
+            Ok(buf.len())
+        })
+    }
+
+    fn read(&self, ep_addr: EndpointAddress, buf: &mut [u8]) -> usb_device::Result<usize> {
+        if !self.is_used(ep_addr) {
+            return Err(UsbError::InvalidEndpoint);
+        }
+
+        if ep_addr.is_in() {
+            return Err(UsbError::InvalidEndpoint);
+        }
+
+        let i = ep_addr.index();
+        interrupt::free(|cs| {
+            let regs = self.periph.borrow(cs);
+
+            // Control EP 0 is special
+            if i == 0 {
+                // Control setup packet is special, since it is put in registers, not a buffer.
+                if regs.events_ep0setup.read().events_ep0setup().bit_is_set() {
+                    regs.events_ep0setup.reset();
+
+                    let ep0_state = unsafe { &mut *self.ep0_state.borrow(cs).as_ptr() };
+
+                    return self.read_control_setup(regs, buf, ep0_state);
+                } else {
+                    // Is the endpoint ready?
+                    if regs.events_ep0datadone.read().events_ep0datadone().bit_is_clear() {
+                        // Not yet ready.
+                        return Err(UsbError::WouldBlock);
+                    }
+                }
+            } else {
+                // Is the endpoint ready?
+                let epdatastatus = regs.epdatastatus.read().bits();
+                if epdatastatus & (1 << (i + 16)) == 0 {
+                    // Not yet ready.
+                    return Err(UsbError::WouldBlock);
+                }
+            }
+
+            // Check that the packet fits into the buffer
+            let size = regs.size.epout[i].read().bits();
+            if size as usize > buf.len() {
+                return Err(UsbError::BufferOverflow);
+            }
+
+            // Clear status
+            if i == 0 {
+                regs.events_ep0datadone.reset();
+            } else {
+                regs.epdatastatus.write(|w| unsafe { w.bits(1 << (i + 16)) });
+            }
+
+            // We checked that the endpoint has data, time to read it
+
+            let epout = [
+                &regs.epout0,
+                &regs.epout1,
+                &regs.epout2,
+                &regs.epout3,
+                &regs.epout4,
+                &regs.epout5,
+                &regs.epout6,
+                &regs.epout7,
+            ];
+            epout[i].ptr.write(|w| unsafe { w.bits(buf.as_ptr() as u32) });
+            // MAXCNT must match SIZE
+            epout[i].maxcnt.write(|w| unsafe { w.bits(size) });
+
+            dma_start();
+            regs.events_endepout[i].reset();
+            regs.tasks_startepout[i].write(|w| w.tasks_startepout().set_bit());
+            while regs.events_endepout[i].read().events_endepout().bit_is_clear() {}
```

Comment:
I suppose that this busy wait works more reliably and should be fine for default buffer sizes.
I also appreciate avoiding an extra buffer copy just to prepare for DMA, so this feels good to me.

Are we concerned about latency/blocking here for larger buffer sizes?

Reply:
I don't know about nrf52 DMA, but on STM32s DMA usually works faster than memcpy. The only way we can optimize this even more is to make this code async (which is not something supported by `usb-device`). Also I don't think nrf52 family has chips with HighSpeed support, so for max 64-byte packet sizes latency/blocking shouldn't be a problem.